XdmodAID.md

```diff
@@ -0,0 +1,265 @@
+
+I then tried to setup Supremm, and the process failed when running from xdmod-setup.
+
+I did get the staging mariadb setup by restoring from a backup of production
+
+#xdmod-hierarchy
+
+What was implemented was a basic history table.  This was done in a simplistic way possible,
+while being mindful of the non-perofromant file system.
+
+I anticipate the need to recreate the hierarchy files some arbitary date.  This could be for
+testing the system with real world data, replicating problems or auditing the system.
+
+The structure is as follows:
+```
+```
+
+In the begining, most of the records will need to be fetched, so the current code just
+selects everything from the table processing everything as it comes, keeping only the
+most recent hierarchy item for each combination of hierachy_id/timestamp.  As this
+SQL statment doens't require a virtual table, it is just reading the table and sending
+it out, we are never writing to the filesystem.
+
+The obvious way to speed this up is to create a table that just has the most recent
+active entires. With such a table, you don't need to do any processing after selecting
+from it, and it is just a select, so this will be a fast operation giving our backend
+filesystem.  Inserting/updating does become more complicated, but this is relatively
+straightforward.
+
+It was suggested that we use the ORM SQLAlchemy.  I considered this, and since it supports
+multiple databases, I would assume that it crafts SQL differetnly baseed on which database
+it is connecting to.  For example, to get good performance from OracleRDB (formerly dec's
+relation database system), queries are written differently than for Oracle.  Although I
+hope it supports a file system that reads qucikly, but doesn't write quickly, I can find
+no guarentee of this, so I have created an issue to explore this in the future when we
+have more actual data (and the backend supports reasonable writing speeds).
+
+It was also suggested that I use a "limit by 1" since all I'm interested in just the most
+recent entry.  Although, on the surface this seems reasonable, there are some problems
+with it.  To pull of the most recent one will require a order implies creating and
+writing a virtual table to perform the sort in and then to return the first record.
+Furthermore, we would have to a query for each hierarchy_id in the hierarchy table.  I
+rejected this as the best way to speed this up is to refactor the code to have an
+aforemented table dedicated to the currently active items.
+
+As output, the xdmod-hierarchy script will produce the following files:
+```
+    hierarchy.csv  - specifies the 3 level hierarcy of institution, field of study, PI
+    group.csv      - has the mapping of nerc project to PIs
+    names.csv      - has the mapping of openshift namespaces and openstack projects to
+                     the nerc project
+    pi2project.csv - maps the openstack project to the openstack project
```

Comment:
openstack project to *nerc* project?

Reply:
The pi2project is needed for the cloud hierarchy.  The way that the xdmod team uses this is to map the openstack project to the name in the names file.  I'll put in an example.

Sparing the confusing details that I could put in, pi2projects.csv is in essence mapping to itself.

Reply:
I have reworked this section, hopefully it is a bit clearer for you.